pkg/pool-stable-phantom/test/StablePoolProtocolFees.test.ts

```diff
@@ -0,0 +1,246 @@
+
+          rateProviders = [];
+          for (let i = 0; i < numberOfTokens; i++) {
+            rateProviders[i] = await deploy('v2-pool-utils/MockRateProvider');
+          }
+
+          if (exemption == Exemption.NONE) {
+            exemptFromYieldProtocolFeeFlags = Array(numberOfTokens).fill(false);
+          } else if (exemption == Exemption.ALL) {
+            exemptFromYieldProtocolFeeFlags = Array(numberOfTokens).fill(true);
+          } else {
+            exemptFromYieldProtocolFeeFlags = range(numberOfTokens).map(() => Math.random() < 0.5);
+
+            if (every(exemptFromYieldProtocolFeeFlags, (flag) => flag == false)) {
+              exemptFromYieldProtocolFeeFlags[0] = true;
+            } else if (every(exemptFromYieldProtocolFeeFlags, (flag) => flag == true)) {
+              exemptFromYieldProtocolFeeFlags[0] = false;
+            }
+          }
+
+          // The rate durations are actually irrelevant since we're forcing cache updates
+          const rateCacheDurations = Array(numberOfTokens).fill(DAY);
+
+          pool = await deploy('MockStablePoolProtocolFees', {
+            args: [
+              vault.address,
+              feesProvider.address,
+              tokens.addresses,
+              rateProviders.map((x) => x.address),
+              rateCacheDurations,
+              exemptFromYieldProtocolFeeFlags,
+            ],
+          });
+        });
+
+        sharedBeforeEach('update rate cache', async () => {
+          // We set new rates for all providers, and then force a cache update for all of them, updating the current
+          // rates. They will now be different from the old rates.
+          await Promise.all(rateProviders.map((provider) => provider.mockRate(fp(random(1.1, 1.5)))));
+          await tokens.asyncEach((token) => pool.updateTokenRateCache(token.address));
+
+          await tokens.asyncEach(async (token) => {
+            const { rate, oldRate } = await pool.getTokenRateCache(token.address);
+            expect(rate).to.not.equal(oldRate);
+          });
+        });
+      }
+
+      function itComputesTheInvariantsCorrectly() {
+        it('computes the swap fee growth invariant correctly', async () => {
+          // The swap fee growth invariant is computed by using old rates for all tokens
+          const oldRateBalances = await Promise.all(
+            balances.map(async (balance, i) => {
+              const { rate, oldRate } = await pool.getTokenRateCache(tokens.get(i).address);
+              return balance.mul(oldRate).div(rate);
+            })
+          );
+
+          const expectedSwapFeeGrowhtInvariant = calculateInvariant(
+            oldRateBalances,
+            AMPLIFICATION_FACTOR.div(AMPLIFICATION_PRECISION)
+          );
+
+          const { swapFeeGrowthInvariant } = await pool.getGrowthInvariants(balances, AMPLIFICATION_FACTOR);
+          expect(swapFeeGrowthInvariant).to.almostEqual(expectedSwapFeeGrowhtInvariant, 1e-10);
```

Comment:
```suggestion
          expect(swapFeeGrowthInvariant).to.almostEqual(expectedSwapFeeGrowthInvariant, 1e-10);
```
Maybe pull the 1e-10 out to a MAX_RELATIVE_ERROR field like in other tests?

Reply:
This is only for the invariants though - other tests will have other errors. i'll look into this later.